diff --git a/src/upf_app/app/upf_config.hpp b/src/upf_app/app/upf_config.hpp
index 3ee7192..f2b0785 100644
--- a/src/upf_app/app/upf_config.hpp
+++ b/src/upf_app/app/upf_config.hpp
@@ -152,7 +152,7 @@ class upf_config {
         pdns(),
         smfs(),
         max_pfcp_sessions(100),
-        nsf(true),
+        nsf({true}),
         enable_snat(false),
         enable_fr(false),
         nrf_addr() {
diff --git a/src/upf_app/simpleswitch/pfcp_pdr.cpp b/src/upf_app/simpleswitch/pfcp_pdr.cpp
index 402f16b..1cb0700 100644
--- a/src/upf_app/simpleswitch/pfcp_pdr.cpp
+++ b/src/upf_app/simpleswitch/pfcp_pdr.cpp
@@ -107,7 +107,7 @@ bool pfcp_pdr::look_up_pack_in_core(

     in_addr _addr4{};
     unsigned char buf_in_addr[sizeof(struct in_addr)];
-    if (inet_pton(AF_INET, "10.0.0.2", buf_in_addr) == 1) {
+    if (inet_pton(AF_INET, connected_ue_ip.c_str(), buf_in_addr) == 1) {
       memcpy(&_addr4, buf_in_addr, sizeof(struct in_addr));
     }

diff --git a/src/upf_app/simpleswitch/pfcp_pdr.hpp b/src/upf_app/simpleswitch/pfcp_pdr.hpp
index 03997b4..ff84abe 100644
--- a/src/upf_app/simpleswitch/pfcp_pdr.hpp
+++ b/src/upf_app/simpleswitch/pfcp_pdr.hpp
@@ -52,6 +52,7 @@ class pfcp_pdr {
   std::pair<bool, pfcp::qer_id_t> qer_id;
   std::pair<bool, pfcp::activate_predefined_rules_t> activate_predefined_rules;
   bool notified_cp;
+  std::string connected_ue_ip;

   explicit pfcp_pdr(uint64_t lseid)
       : lock(),
diff --git a/src/upf_app/simpleswitch/pfcp_switch.cpp b/src/upf_app/simpleswitch/pfcp_switch.cpp
index 7cb85f4..e204325 100644
--- a/src/upf_app/simpleswitch/pfcp_switch.cpp
+++ b/src/upf_app/simpleswitch/pfcp_switch.cpp
@@ -897,6 +897,22 @@ void pfcp_switch::handle_pfcp_session_establishment_request(
               << std::endl;
     for (const auto& it : up_seid2pfcp_sessions) {
       std::cout << it.second->to_string() << std::endl;
+      for (std::vector<std::shared_ptr<pfcp::pfcp_pdr>>::const_iterator it_pdr =
+              it.second->pdrs.begin();
+          it_pdr != it.second->pdrs.end(); ++it_pdr) {
+        std::shared_ptr<pfcp::pfcp_pdr> pdr = *it_pdr;
+
+        if (pdr->pdi.first) {
+          if (pdr->pdi.second.ue_ip_address.first) {
+            std::string ip = {};
+            if (pdr->pdi.second.ue_ip_address.second.v4) {
+              ip = oai::utils::conv::toString(
+                  pdr->pdi.second.ue_ip_address.second.ipv4_address);
+              connected_ue_ip = ip;
+            }
+          }
+        }
+      }
     }
   }
 }
@@ -1360,7 +1376,7 @@ void pfcp_switch::pfcp_session_look_up_pack_in_core(
     // Set UE IP addr to the adress of the Relay
     in_addr _addr4{};
     unsigned char buf_in_addr[sizeof(struct in_addr)];
-    if (inet_pton(AF_INET, "10.0.0.2", buf_in_addr) == 1) {
+    if (inet_pton(AF_INET, connected_ue_ip.c_str(), buf_in_addr) == 1) {
       memcpy(&_addr4, buf_in_addr, sizeof(struct in_addr));
     }
     ue_ip = ntohl(_addr4.s_addr);
@@ -1377,6 +1393,7 @@ void pfcp_switch::pfcp_session_look_up_pack_in_core(
       for (std::vector<std::shared_ptr<pfcp::pfcp_pdr>>::iterator it =
                pdrs->begin();
            it < pdrs->end(); ++it) {
+        (*it)->connected_ue_ip = connected_ue_ip;
         if ((*it)->look_up_pack_in_core(iph, num_bytes)) {
           std::shared_ptr<pfcp::pfcp_session> ssession = {};
           uint64_t lseid                               = 0;
diff --git a/src/upf_app/simpleswitch/pfcp_switch.hpp b/src/upf_app/simpleswitch/pfcp_switch.hpp
index adbb9b1..8432b62 100644
--- a/src/upf_app/simpleswitch/pfcp_switch.hpp
+++ b/src/upf_app/simpleswitch/pfcp_switch.hpp
@@ -74,6 +74,7 @@ class pfcp_switch {
   folly::MPMCQueue<iovec_q_item_t*>* free_pool_;
   folly::MPMCQueue<iovec_q_item_t*>* work_pool_;
   uint32_t num_threads_;
+  std::string connected_ue_ip;
   char* recv_buffer_alloc_;
   char recv_buffer_[PFCP_SWITCH_RECV_BUFFER_SIZE];
   std::thread pwThread_;
